### ISSUE TITLE
Pods keep failing with FailedCreatePodSandBox

### ISSUE THREAD

**Issue by nielsvz** [cilium/cilium#26777]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We have two clusters that are dedicated to running GitLab CI and keep getting `FailedCreatePodSandBox` every couple of days, the issue goes away for a while when we restart Cilium.

The output of `cilium status` shows one error, the endpoint is listed as `not-ready`, and we see the following events for the specific Pods with this issue:

```
root@ip-10-210-54-51:/home/cilium# cilium-health get   
Daemon uptime:          25h54m29.737214151s
Node load:              0.62 0.60 0.84
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.24+ (v1.24.14-eks-c12679a) [linux/amd64]
Kubernetes APIs:        ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Disabled   
Host firewall:          Disabled
CNI Chaining:           aws-cni
CNI Config file:        CNI configuration file management disabled
Cilium:                 Ok   1.13.4 (v1.13.4-4061cdfc)
NodeMonitor:            Listening for events on 8 CPUs with 64x4096 of shared memory
Cilium health daemon:   Ok   
IPAM:                   IPv4: 1/254 allocated from 100.127.0.0/24, 
IPv6 BIG TCP:           Disabled
BandwidthManager:       Disabled
Host Routing:           Legacy
Masquerading:           Disabled
Controller Status:      84/85 healthy
  Name                                  Last success   Last error   Count   Message
  endpoint-2787-regeneration-recovery   never          5s ago       2       regeneration recovery failed   
Proxy Status:            OK, ip 100.127.0.236, 1 redirects active on ports 10000-20000
Global Identity Range:   min 256, max 65535
Hubble:                  Ok   Current/Max Flows: 4095/4095 (100.00%), Flows/s: 32.18   Metrics: Disabled
Encryption:              Disabled
```

```
Warning  FailedCreatePodSandBox  2m15s                   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "6040f000bbe39140ac313ed8305999de2d9ca8ca4a285328558126ff8a1cbb60": plugin type="cilium-cni" name="cilium" failed (add): unable to create endpoint: Cilium API client timeout exceeded                                                                                   
Warning  FailedCreatePodSandBox  31s                     kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "aaad9a0f400b3740451eaf4a7ac1f5b12636f60991723cc36871095ccab8b4b3": plugin type="cilium-cni" name="cilium" failed (add): unable to create endpoint: Cilium API client timeout exceeded                                                                                   
Warning  FailedCreatePodSandBox  2s                      kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "97a26f8a313d9950d7cf4933f4096400f0adac7ca97e57d69bdf704f30f94648": plugin type="cilium-cni" name="cilium" failed (add): unable to create endpoint: Put "http:///var/run/cilium/cilium.sock/v1/endpoint/cilium-local:0": EOF   
```

### Cilium Version

v1.13.4

Also observed in v1.12.4 and v1.12.9

### Kernel Version

5.15.108 (EKS Bottlerocket)

### Kubernetes Version

1.24.14

### Sysdump

I am not able to share the sysdump here because it may contain sensitive details.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by nielsvz** [comment]
I believe this issue is caused by the PolicyMap limits. We have a number of Pod labels that include commit refs, branch names and job IDs that are only used for reporting purposes, but not used by Network Policies at all.

We've prefixed the unused labels with `annotation.` to make Cilium ignore them as documented [here](https://docs.cilium.io/en/stable/operations/performance/scalability/identity-relevant-labels/#limiting-identity-relevant-labels)

**Comment by nielsvz** [comment]
After modifying pod labels to exclude unique values we've noticed the issue still occurs, but less frequently. I have two separate sysdumps I could share privately.

**Comment by christarazi** [comment]
Could you provide more details about the crash like what sort of changes happen in the cluster before it crashes? What sort of policies are being used, toCIDRs, toFQDNs, etc, how are you applying policies to that one problematic endpoint?

**Comment by nielsvz** [comment]
We have several clusters dedicated to GitLab CI (using their Kubernetes executor). The following policies are applied to all Pods created by GitLab:
1. deny-all for both Ingress and Egress traffic (`NetworkPolicy`)
1. allow-cluster-dns for Egress to coredns in kube-system (`CiliumNetworkPolicy`)
1. specific policies that control Egress traffic from Pods (GitLab CI Jobs) running there based on certain variables exposed from GitLab (group/ownership details), policies select these pods in one of two ways:
    1. `endpointSelector` with `matchLabels` and exactly one label with a specific value (example: `gitlab.acme.org/project-id: 1234`
    1. `endpointSelector` with `matchExpressions` and exactly one label with multiple values (expression operator is `In`)

Almost all policies use `toFQDNs` with either `matchName` or `matchPattern`

At some point (hours to days), we notice all new Pods generated by GitLab CI fail to start with the `FailedCreatePodSandBox` event, contained on one host, eventually all hosts get the same issue. When we restart the Cilium agent on the affected host, the issue is resolved (immediately after we see it regenerate endpoints the Pods will become Ready).

**Comment by christarazi** [comment]
Thanks for clarifying.

When Cilium is near crashing or just crashed, could you share the following:

1. `cilium metrics list` output (ideally before crash; after crash, the metrics will reset which is not useful)
2. the Cilium logs so we can understand the cause of the crash

**Comment by nielsvz** [comment]
Attached you'll find the output of `cilium metrics list` and `kubectl logs` from one of our clusters, at the time of extracting the metrics and logs only one node was failing Pods with the `FailedCreatePodSandBox` Event, usually the other nodes follow soon after that as well.

[cilium-metrics-logs.zip](https://github.com/cilium/cilium/files/12212421/cilium-metrics-logs.zip)
This contains a .txt file for the `cilium metrics list` output and a .log file for `kubectl logs` for each of the nodes in the cluster.

**Comment by kbaegis** [comment]
I am also seeing this.

EKS: 1.27
VPC CNI: v1.13.3-eksbuild.1
CoreDNS: v1.10.1-eksbuild.2
kube-proxy: v1.27.3-eksbuild.2
worker: v1.27.3-eks-a5565ad
```
# cilium status
    /¯¯\
 /¯¯\__/¯¯\    Cilium:             OK
 \__/¯¯\__/    Operator:           OK
 /¯¯\__/¯¯\    Envoy DaemonSet:    disabled (using embedded mode)
 \__/¯¯\__/    Hubble Relay:       OK
    \__/       ClusterMesh:        disabled

Deployment             cilium-operator    Desired: 1, Ready: 1/1, Available: 1/1
Deployment             hubble-ui          Desired: 1, Ready: 1/1, Available: 1/1
Deployment             hubble-relay       Desired: 1, Ready: 1/1, Available: 1/1
DaemonSet              cilium             Desired: 2, Ready: 2/2, Available: 2/2
Containers:            cilium             Running: 2
                       cilium-operator    Running: 1
                       hubble-ui          Running: 1
                       hubble-relay       Running: 1
Cluster Pods:          0/6 managed by Cilium
Helm chart version:    1.13.4
Image versions         cilium             quay.io/cilium/cilium:v1.13.4@sha256:bde8800d61aaad8b8451b10e247ac7bdeb7af187bb698f83d40ad75a38c1ee6b: 2
                       cilium-operator    quay.io/cilium/operator-aws:v1.13.4@sha256:c6bde19bbfe1483577f9ef375ff6de19402ac20277c451fe05729fcb9bc02a84: 1
                       hubble-ui          quay.io/cilium/hubble-ui:v0.11.0@sha256:bcb369c47cada2d4257d63d3749f7f87c91dde32e010b223597306de95d1ecc8: 1
                       hubble-ui          quay.io/cilium/hubble-ui-backend:v0.11.0@sha256:14c04d11f78da5c363f88592abae8d2ecee3cbe009f443ef11df6ac5f692d839: 1
                       hubble-relay       quay.io/cilium/hubble-relay:v1.13.4@sha256:bac057a5130cf75adf5bc363292b1f2642c0c460ac9ff018fcae3daf64873871: 1
```

**Comment by christarazi** [comment]
> We have several clusters dedicated to GitLab CI (using their Kubernetes executor). The following policies are applied to all Pods created by GitLab:
> 
>     1. deny-all for both Ingress and Egress traffic (`NetworkPolicy`)
> 
>     2. allow-cluster-dns for Egress to coredns in kube-system (`CiliumNetworkPolicy`)
> 
>     3. specific policies that control Egress traffic from Pods (GitLab CI Jobs) running there based on certain variables exposed from GitLab (group/ownership details), policies select these pods in one of two ways:
>        
>        1. `endpointSelector` with `matchLabels` and exactly one label with a specific value (example: `gitlab.acme.org/project-id: 1234`
>        2. `endpointSelector` with `matchExpressions` and exactly one label with multiple values (expression operator is `In`)
> 
> 
> Almost all policies use `toFQDNs` with either `matchName` or `matchPattern`
> 
> At some point (hours to days), we notice all new Pods generated by GitLab CI fail to start with the `FailedCreatePodSandBox` event, contained on one host, eventually all hosts get the same issue. When we restart the Cilium agent on the affected host, the issue is resolved (immediately after we see it regenerate endpoints the Pods will become Ready).

Without a sysdump, it's hard to say whether Cilium is behaving as expected / configured or whether there's a look. AFAICT, before the crash, there are many, many local identities (`node_local`; identities backed by a CIDR, i.e. FQDN as well):

```
$ rg cilium_identity
cilium-agent-healthy-2.txt
135:cilium_identity                                       type="cluster_local"                                                     17.000000
136:cilium_identity                                       type="node_local"                                                        2887.000000
137:cilium_identity                                       type="reserved"                                                          8.000000

cilium-agent-healthy-1.txt
130:cilium_identity                                       type="cluster_local"                                                     12.000000
131:cilium_identity                                       type="node_local"                                                        1210.000000
132:cilium_identity                                       type="reserved"                                                          8.000000

cilium-agent-failed-before-restart.txt
139:cilium_identity                                       type="cluster_local"                                                     18.000000
140:cilium_identity                                       type="node_local"                                                        6448.000000
141:cilium_identity                                       type="reserved"                                                          8.000000

cilium-agent-failed-after-restart.txt
121:cilium_identity                                       type="cluster_local"                                                     17.000000
122:cilium_identity                                       type="node_local"                                                        6453.000000
123:cilium_identity                                       type="reserved"                                                          8.000000
```

The configured and default policymap size is 

```
level=info msg="  --bpf-policy-map-max='16384'" subsys=daemon
```

so my first guess is that given the policies installed + the number of node-local identities, the policymap is overflowing due to the sheer number of identities in use by the endpoints.

I'd recommend looking into FQDN side and why there are so many active IPs/identities, before just bumping the policymap max. Bumping the max without understanding the cause risks kicking the can down the line. One hint is that the default min TTL for each FQDN is configured at `1h`,  so try out `--tofqdns-min-ttl=0s` to use the default TTL value of the FQDN. This should cause Cilium to GC identities for quicker, rather than keep them for a default of `1h`. Keep in mind that Cilium will keep an IP <-> identity mapping alive for as long as the IP is actively used (this is done via conntrack bookkeeping).

**Comment by nielsvz** [comment]
I've managed to remove some of the sensitive data we're not allowed to share from these dumps:

[cilium-sysdump-tf-gitlab-ci-1-20230713.zip](https://github.com/cilium/cilium/files/12289159/cilium-sysdump-tf-gitlab-ci-1-20230713.zip)
[cilium-sysdump-tf-gitlab-ci-1-20230714.zip](https://github.com/cilium/cilium/files/12289160/cilium-sysdump-tf-gitlab-ci-1-20230714.zip)

**Comment by nielsvz** [comment]
It looks like the `--tofqdns-min-ttl` defaults to 0 since v1.14.0: https://github.com/cilium/cilium/pull/21439

We'll test if upgrading to v1.14.0 will resolve this issue for us.

**Comment by christarazi** [comment]
@nielsvz Just a heads up that there's a regression in v1.14.0 where after restarting Cilium on that version, any IP <-> identity mappings for CIDRs or FQDNs are wiped out which could cause connectivity drops. The fix has been merged and is slated for v1.14.1: https://github.com/cilium/cilium/pull/27327

**Comment by nielsvz** [comment]
We've upgraded one of our CI clusters to Cilium v1.14 with `--tofqdns-min-ttl` set to the default value (0) and have not observed any `FailedCreatePodSandBox` Events since.

I believe this issue was caused because we run many CI jobs that connect to AWS API endpoints, those seem to have a 60 second TTL and can return different IPs every minute. Combined with the old default minimum TTL of 3600 seconds I believe this was creating too many policy map entries for Cilium, eventually exceeding the limit of 16384 entries.
